src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,1015 @@
+        elif typecheck(val, fmpz_mpoly):
+            if ctx is None:
+                ctx = fmpq_mpoly_ctx.from_context((<fmpz_mpoly>val).ctx)
+            elif not typecheck(ctx, fmpq_mpoly_ctx) and not typecheck(ctx, fmpz_mpoly_ctx):
+                raise TypeError(f"{ctx} is not a fmpq_mpoly_ctx or fmpz_mpoly_ctx")
+            elif ctx.nvars() != val.context().nvars():
+                raise ValueError(
+                    f"Provided context ('{ctx}') and provided fmpz_mpoly ('{val}') don't share the same number of variables"
+                )
+            init_fmpq_mpoly(self, ctx)
+            fmpz_mpoly_set(self.val.zpoly, (<fmpz_mpoly>val).val, (<fmpz_mpoly> val).ctx.val)
+            fmpq_one(self.val.content)
+            fmpq_mpoly_reduce(self.val, self.ctx.val)
+        elif isinstance(val, dict):
+            if ctx is None:
+                raise ValueError("A context is required to create a fmpq_mpoly from a dict")
+            x = ctx.from_dict(val)
+            # XXX: this copy is silly, have a ctx function that assigns an fmpz_mpoly_t
+            init_fmpq_mpoly(self, ctx)
+            fmpq_mpoly_set(self.val, (<fmpq_mpoly>x).val, self.ctx.val)
+        elif isinstance(val, str):
+            if ctx is None:
+                raise ValueError("Cannot parse a polynomial without context")
+            val = val.encode("ascii")
+            init_fmpq_mpoly(self, ctx)
+            fmpq_mpoly_set_str_pretty(self.val, val, self.ctx.c_names, self.ctx.val)
+            fmpq_mpoly_sort_terms(self.val, self.ctx.val)
+        else:
+            v = any_as_fmpq(val)
+            if v is NotImplemented:
+                raise TypeError("cannot create fmpz_mpoly from type %s" % type(val))
+            if ctx is None:
+                raise ValueError("Need context to convert  fmpz to fmpq_mpoly")
+            init_fmpq_mpoly(self, ctx)
+            fmpq_mpoly_set_fmpq(self.val, (<fmpq>v).val, self.ctx.val)
+
+    def __bool__(self):
+        return not fmpq_mpoly_is_zero(self.val, self.ctx.val)
+
+    def __richcmp__(self, other, int op):
+        if not (op == Py_EQ or op == Py_NE):
+            return NotImplemented
+        elif other is None:
+            return op == Py_NE
+        elif typecheck(self, fmpq_mpoly) and typecheck(other, fmpq_mpoly):
+            if (<fmpq_mpoly>self).ctx is (<fmpq_mpoly>other).ctx:
+                return (op == Py_NE) ^ bool(
+                    fmpq_mpoly_equal((<fmpq_mpoly>self).val, (<fmpq_mpoly>other).val, (<fmpq_mpoly>self).ctx.val)
+                )
+            else:
+                return op == Py_NE
+        else:
+            return NotImplemented
+
+    def __len__(self):
+        return fmpq_mpoly_length(self.val, self.ctx.val)
+
+    def __getitem__(self, x):
+        """
+        Return the term at index `x` if `x` is an `int`, or the term with the exponent
+        vector `x` if `x` is a tuple of `int`s or `fmpq`s.
```

Comment:
Defining both `__len__` and `__getitem__` implicitly makes the object iterable:
```python
In [7]: p = fmpq_mpoly('x0*2 + x1*3', ctx)

In [8]: p
Out[8]: 2*x0 + 3*x1

In [9]: list(p)
Out[9]: [2*x0, 3*x1]
```
We should consider what is really the intended behaviour here because the polynomial behaves sort of like a sequence and sort of like a dict.

For dict-like behaviour there should be the equivalent of `.keys`, `.values` and `.items`.

Also `__contains__` is not defined but will implicitly use iterability (likely not efficient):
```python
In [14]: p2
Out[14]: 3*x1

In [15]: p
Out[15]: 2*x0 + 3*x1

In [16]: p2 in p
Out[16]: True

In [17]: 2*p2 in p
Out[17]: False
```
Again I'm not sure what should really be the expected behaviour here. Conceptually I think that overloading `__getitem__` like this will be confusing for users so it should be made clear what sort of Python container-like interface this implements.

Reply:
This is inconsistent for sequence like behaviour:
```python
In [30]: p[1] = p[1]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[30], line 1
----> 1 p[1] = p[1]

File fmpq_mpoly.pyx:317, in flint.types.fmpq_mpoly.fmpq_mpoly.__setitem__()

TypeError: provided coefficient not coercible to fmpq
```

Reply:
In general I think it is better to use explicit methods rather than overloading operators unless it is conceptually clear what interface is being implemented with the operators.

Reply:
> For dict-like behaviour there should be the equivalent of .keys, .values and .items.

Given `fmp[zq]_mpoly`s are constructed from dictionaries I think it's reasonable for them to behave like one, with the keys being the exponent vector, values as the coefficients, and items the monomials (?). 

The interface I'd implemented previously with `__getitem__` and `__setitem__` absolutely needs to be changed. To be consistent with univariate polynomials I'll change the `__getitem__` and `__setitem__` methods to index only on term position and operate with coefficients.

However, a polynomial isn't a dictionary and cannot behave exactly like one. I think a reasonable based on a dict might have
- `list(p)` return a list of exponent vectors like a dict
- `p[idx]` return the coefficient of the `idx`th term, like univariate polynomials. This isn't consistent with a dict
- `p[idx] = c` sets the coefficient of the `idx`th term, like univariate polynomials. This isn't consistent with a dict
- `__contains__` should probably operate on terms, not exponent vectors. This isn't consistent with a dict
- `get()` return a term (?) based on an exponent vector
- `items()` return an iterable of terms, perhaps this could be made more dict like be returning a tuple of `coeff, exp vec`
- `keys()` return an iterable of exponent vectors
- `values()` return an iterable of coefficients

This is essentially a dictionary but with indexing replaced with typical sequence indexing.
I don't think the other dictionary operations like `del`, `pop`, etc are of any use here.

Reply:
> Given `fmp[zq]_mpoly`s are constructed from dictionaries I think it's reasonable for them to behave like one, with the keys being the exponent vector, values as the coefficients, and items the monomials (?).

It might be worth looking at the analogous types in SymPy which literally are a dict subclass:
```python
In [22]: from sympy import *

In [23]: x = symbols('x')

In [24]: R = QQ[x]

In [25]: p = R(x**2 - 1)

In [26]: p
Out[26]: x**2 - 1

In [27]: dict(p)
Out[27]: {(0,): -1, (2,): 1}

In [28]: list(p)
Out[28]: [(0,), (2,)]

In [30]: type(p)
Out[30]: sympy.polys.rings.PolyElement

In [31]: type(p).mro()
Out[31]: 
[sympy.polys.rings.PolyElement,
 sympy.polys.rings.PolyElement,
 sympy.polys.domains.domainelement.DomainElement,
 sympy.printing.defaults.Printable,
 sympy.core.sympify.CantSympify,
 dict,
 object]
```
Note that in SymPy these are treated as immutable.

Reply:
I've just pushed some changes that make `fmpz_mpoly` and `fmpq_mpoly` more dict-like. This includes implementing `.keys()`, `.values()`, `.items()`, `__contains__`, and `__iter__`. As well as reverting to exponent vector indexing.

Reply:
The SymPy versions alias methods like keys and values with more polynomial sounding names:
```python
In [2]: p = QQ[x](x**2 + 2*x + 3)

In [3]: p.monoms()
Out[3]: [(2,), (1,), (0,)]

In [4]: p.coeffs()
Out[4]: [1, 2, 3]

In [5]: p.terms()
Out[5]: [((2,), 1), ((1,), 2), ((0,), 3)]
```
There are also analogous iter methods that more directly return the dict iterators:
```python
In [6]: p.itermonoms()
Out[6]: <dict_keyiterator at 0x7556b7c63380>
```
Using `.keys` etc seems okay to me though.

Reply:
That's quite reasonable, I'll make those aliases tonight.